Migration options api to composition api

NicolasGujer opened the pull request:
1. Refactored from the Options-API to the Composition-API.
2. Replaced the storage.ts with a Pinia store for state management.

NicolasGujer committed:
--> Migration from options-api to composition-api in all midata components and pages, loginCard, PageLinks, Overview and Error404

--> removed unused midata components

NicolasGujer committed:
--> Fixed dead links
--> Migrated all components and pages to Composition-Api
--> Because <setup> doesnt support global properties, they need to be imported every time they are used. In the <template> they can still be used
--> Internationalization.vue now uses only the 3 supported locales. Switching is consistent across the whole page. This change may be reverted if it is a requirement to set date locale interdependently from language locale

NicolasGujer committed:
Changes:
--> Upgraded to the newest Version of Quasar for proper Pinia support (quasar upgrade && quasar upgrade -i)
--> Migration from storage.ts to user.ts pinia store

NicolasGujer committed:
Changes:
--> FilteredList is now a computed property in the user pinia store. The code-filter can be changed with the "currentFilter" ref
--> Removed "restoreFromStorage" because the values are automatically saved in sessionStorage
--> the "changeLanguage" function is now a pinia store function
--> The expansion of the patientResource and the ObservationList is now saved in the store and controlled from there. The user preference is saved even on page reload

Commit:
changes:
--> Added 'mdi-v6' as Icon library (pineapple icon needed)
--> Renamed some functions in the UserStore and converted all to async/await
--> Uniform locale in the whole App ('en-gb', 'fr-ch', 'de-ch')
--> Creating an Observation doesnt load all Observations from midata anymore. Instead it just pushes the newly created Observation to the array
--> New Pinia and Vue Devtools Page

Commit:
changes:
--> Fixed a small visual bugg in code demo
--> Added instructions on how to open the developer menu in chrome and firefox
--> Removed console.log statements with no usage

Commit:
changes:
--> Removed the storage.ts plugin class (the pinia store replaced the functionality).
--> Fixed a bugg in the PatientResource Component. It was still dependent on the storage.ts plugin.

heg2 commented on src/pages/midata/MidataDemo.vue:
```diff
@@ -52,7 +46,7 @@
     <div style="height: 25px"></div>
 
-    <PatientResource :flag='flag' :patient-resource='patientResource'></PatientResource>
+    <PatientResource :visible='flag'></PatientResource>
```
Eventuell übersehe ich etwas, aber die Sichtbarkeit der Komponente würde ich direkt hier über v-if lösen, statt einem Property, das dann innerhalb der Komponente die Sichtbarkeit ändert (klar, war schon vor dem Refactoring so, aber ich würde es trotzdem ändern)

heg2 commented on src/components/midata/PatientResource.vue:
```diff
@@ -1,6 +1,6 @@
 <template>
-  <q-card v-if="flag">
-    <q-card-section v-if="!isEmpty(patientResource)">
+  <q-card v-if="visible">
```
Siehe Kommentar bei MidataDemo.vue - ich würde das v-if ausserhalb der Komponente regeln

heg2 commented on src/pages/development/DevelopmentBasics.vue:
```diff
@@ -101,7 +101,7 @@
         type="a"
         target="_blank"
-        :href="'https://nodejs.dev/download/'"
+        :href="'https://nodejs.dev/en/download/'"
```
sehr aufmerksam 👍

heg2 commented on src/pages/development/PiniaAndDevtools.vue:
```diff
@@ -0,0 +1,220 @@
+        Options-API funktionieren. Mit Pinia kann man sogenannte "stores" erstellen, die dann thematisch zusammenhängenden
+        State beinhalten. Man könnte auch den gesamten state in einem store verwalten, aber bei grösseren Projekten
+        liegt es nahe, diesen aufzuteilen. Diese stores können in den Komponenten und Pages genutzt werden. Alle Veränderungen
+        dieses stores werden automatisch in allen anderen Komponenten angezeigt. Weitere Infos und Anleitungen finden sie in
```
"Sie" grossgeschrieben

heg2 commented on src/pages/development/PiniaAndDevtools.vue:
```diff
@@ -0,0 +1,220 @@
+    <div class="q-my-xl">
+      <div class="q-qb-sm text-h5">Was ist State Management?</div>
+      <p>
+        State Management beschreibt das Verwalten des Zustands eines oder mehreren User-Interfaces (UI) einer Applikation.
```
In den folgenden Absätzen sollten "Tool", "State", "Store" etc. konsequent entweder gross oder klein geschrieben werden. Ich wäre für gross, da es besser in den Lesefluss passt im deutschen Text.

heg2 commented on src/pages/development/PiniaAndDevtools.vue:
```diff
@@ -0,0 +1,220 @@
+    <div class="q-my-xl">
+      <div class="q-qb-sm text-h5">Pinia in dieser Applikation</div>
+      <p>
+        Pinia wird in dieser Applikation genutzt, um den Zustand rund um die Midata Demo zu verwalten. Alle Observationen
```
"um den Zustand der MIDATA-Demo zu verwalten", oder?

heg2 commented on src/pages/development/PiniaAndDevtools.vue:
```diff
@@ -0,0 +1,220 @@
+      <div class="q-qb-sm text-h5">Pinia in dieser Applikation</div>
+      <p>
+        Pinia wird in dieser Applikation genutzt, um den Zustand rund um die Midata Demo zu verwalten. Alle Observationen
+        und die Patienten-Ressource von Midata sind in einem Pinia store gespeichert und werden dort verwaltet.
```
MIDATA in All-Caps zwecks Konsistenz

heg2 commented on src/pages/development/PiniaAndDevtools.vue:
```diff
@@ -0,0 +1,220 @@
+        Pinia wird in dieser Applikation genutzt, um den Zustand rund um die Midata Demo zu verwalten. Alle Observationen
+        und die Patienten-Ressource von Midata sind in einem Pinia store gespeichert und werden dort verwaltet.
+        Auch Präferenzen des Nutzers wie die ausgewählte Sprache <a href="#/internationalization">
+        (mehr dazu auf der Internationalization Seite)</a> und ob die kompletten Observationen sichtbar sein sollten oder nicht,
```
Link zur I18N Seite einfügen

heg2 reviewed (commented):
Sieht grundsätzlich gut aus, danke!

Was man noch machen könnte, wenn wir schon beim Refactoren sind: 
- [moment.js ist quasi deprecated](https://momentjs.com/docs/#/-project-status/), man könnte es vermutlich durch [das native JS Intl](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl) ersetzen)
- npm audit hat noch drei Vulnerabilities, die könnten wir auch noch versuchen zu fixen (evtl. reicht bereits `npm audit fix
- was mir noch aufgefallen ist (war aber schon vor dem Refactoring so): Wenn man in der MIDATA-Demo eine der Ressourcen bearbeitet, sind im Popup nicht die bisherigen Werte der Ressource geladen, sondern die Default-Werte für Blutdruck oder Temperatur. Das könnte man auch noch fixen

AngryBacteria committed:
changes:
--> fixed vulnerabilities with "npm audit fix"
--> The edit dialog now renders with the current values of the Observation
--> The visibility of the PatientResource component is now handled with the userStore. If there is no resource the component wont be visible
--> Fixed type-errors in the Pinia page
--> started to convert from momentjs to the Javascript Intl API

Commit:
changes:
--> replaced momentjs with intl

Commit:
quasar virtual-scroll issue fixed
--> The quasar virtual scroll had an issue while scrolling. It jumped to the top constantly when the items array was bigger than 40 elements. That is fixed now by wrapping all the elements inside the v-slot into a div. Thanks heg2 for the fix idea.

Commit:
Small fixes
--> Removed references to the removed momentjs package in the pages Internationalization.vue and Quasar.vue and the boot file plugins.ts
--> Fixed some inconsistencies in the documentation

heg2 committed:
fix typo

heg2 commented:
tiptop 👍
